### PR TITLE
Feature/inhibit if kmods are not supported

### PR DESCRIPTION
**Pull request by ZhukovGreen**
~~**blocked by:** #191 (rebased on top)~~

This MR introduces a check of host kernel modules against available kernel modules on rhel

TODO:
- [x] verify 
- [x] Move 3rd party kmods check right after system facts check
- [x] Use repoqery -l {pkgs} instead of loop for each package

### CODE REVIEW

**Comment by SpyTec** [comment]
Is this ready for review @ZhukovGreen?

**Comment by ZhukovGreen** [comment]
@SpyTec Not yet Eric

**Comment by shaded-enmity** [comment]
The code should also check currently loaded modules, something like:
```bash
lsmod | cut -d' ' -f1 | tail -n +2 | sort -R | xargs modinfo | grep '^filename' | tr -s ' ' | cut -d' ' -f2
```

The main rationale here is that if I'm using non-standard kmods, more than likely they live in non-standard locations.

**Comment by ZhukovGreen** [comment]
@shaded-enmity but even custom kernel modules should live in /lib/modules/{kernel-version}. This is considered in the implementation

**Comment by shaded-enmity** [comment]
@ZhukovGreen It is a convention that all kmods in EL live in these folders, yes. But there's nothing preventing me from placing my kmods in, say, `/opt/myCompany/myKmods`, kernel doesn't really care from what directory it loads these mods. A lot of 3rd parties do exactly this -- place everything under `/opt/myCompany`.

If we want to separate the implementation into two parts I'm fine with that:
- Kernel module files installed in standard location (this PR)
- Kernel modules actively loaded on the system from *any* location (my suggestion)

This is all a matter of setting the right expectations for QE and support folks.

**Comment by ZhukovGreen** [comment]
@shaded-enmity yep, I would rather split it into two parts. Based on sync with the CEE guys, they are ok if we cover standard location first

File: convert2rhel/unit_tests/conftest.py
Comment (SpyTec) [inline convert2rhel/unit_tests/conftest.py:77]: Missing doc

File: convert2rhel/checks.py
Comment (SpyTec) [inline convert2rhel/checks.py:184]:
```suggestion
            "Unsigned kernel module(s) has been detected. "
            "Third-party components are not supported per our "
            "software support policy\n%s.\n\n "
            "Uninstall or disable the module(s) and run convert2rhel "
            "again to continue with the conversion.",
```

File: convert2rhel/checks.py
Comment (SpyTec) [inline convert2rhel/checks.py:196]:
```suggestion
            "The following kernel module(s) are not supported "
```

File: convert2rhel/unit_tests/checks_test.py
Comment (SpyTec) [inline convert2rhel/unit_tests/checks_test.py:366]: Out of scope for next release
```suggestion
```

Also has a typo in the method name

File: convert2rhel/checks.py
Comment (SpyTec) [inline convert2rhel/checks.py:127]: Could use more inline comments to explain what some pieces of code does

File: convert2rhel/unit_tests/checks_test.py
Comment (SpyTec) [inline convert2rhel/unit_tests/checks_test.py:109]: Tests shouldn't be testing two methods at the same time, should split into two tests

File: convert2rhel/checks.py
Comment (SpyTec) [inline convert2rhel/checks.py:112]:
```suggestion
    When we scan kernel modules provided by kernel packages
    it is expensive to check each kernel pkg. Since each new
    kernel pkg do not deprecate kernel modules we only select
    the most recent ones.
```

File: convert2rhel/toolopts.py
Comment (SpyTec) [inline convert2rhel/toolopts.py:105]: Out of scope
```suggestion
```

File: convert2rhel/checks.py
Comment (SpyTec) [inline convert2rhel/checks.py:234]: Out of scope
```suggestion
    ensure_compatibility_of_kmods()
```

**Review by SpyTec — CHANGES_REQUESTED** [review]
Overall implementation seems sound by just looking through the code, albeit a bit complex at parts. There's instances where a new variable could describe better what the returned list is as opposed to doing it inline. 

Allowing the user to overwrite the kmods inhibitor is out of scope for the next release, should be removed before merging. That's the only concern with the implementation - the rest of my comments is with the tests, some notes for docs and logging

I have not tested this PR manually yet

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py]: Unexpected package:
kmod-ssss:wrong-version.no-el8.no-cpu-arch
is a source of kernel modules.

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py]: thx

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py:127]: yeap

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py]: Just removed this, do not create confusion. I'll create a jira ticket as a follow up, which will contain this info

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py]: thx

**Review by ZhukovGreen — COMMENTED** [review]
@SpyTec  thx for your review, all comments except one is addressed

File: convert2rhel/unit_tests/checks_test.py
Comment (ZhukovGreen) [reply convert2rhel/unit_tests/checks_test.py:146]: Let me leave my impl. It is important to keep if host_kmods == HOST_MODULES_STUB_GOOD

File: convert2rhel/unit_tests/conftest.py
Comment (ZhukovGreen) [reply convert2rhel/unit_tests/conftest.py]: just removed it, and made private. it is not assumed to be used outside of the pretend_centos8

File: convert2rhel/unit_tests/conftest.py
Comment (ZhukovGreen) [reply convert2rhel/unit_tests/conftest.py:86]: yes

File: convert2rhel/unit_tests/main_test.py
Comment (ZhukovGreen) [reply convert2rhel/unit_tests/main_test.py]: thx

**Review by ZhukovGreen — COMMENTED** [review]
@SpyTec added another set of fixes

**Comment by shaded-enmity** [comment]
Test results from CentOS7 with custom kmod (both installed and just present on the FS)
```
# cat main.c
#include <linux/module.h>	/* Needed by all modules */
#include <linux/kernel.h>	/* Needed for KERN_INFO */

int init_module(void)
{
	printk(KERN_INFO "Hello world 1.\n");
	return 0;
}

void cleanup_module(void)
{
	printk(KERN_INFO "Goodbye world 1.\n");
}

# ls -lah
total 224
-rw-r--r--. 1 root root    256 Mar 22 09:30 main.c
-rw-r--r--. 1 root root 100992 Mar 22 09:36 main.ko
-rw-r--r--. 1 root root    907 Mar 22 09:36 main.mod.c
-rw-r--r--. 1 root root  58856 Mar 22 09:36 main.mod.o
-rw-r--r--. 1 root root  45072 Mar 22 09:36 main.o
-rw-r--r--. 1 root root    172 Mar 22 09:36 Makefile
-rw-r--r--. 1 root root     26 Mar 22 09:36 modules.order
-rw-r--r--. 1 root root      0 Mar 22 09:36 Module.symvers

# modinfo main.ko
filename:       /root/kmod/main.ko
retpoline:      Y
rhelversion:    7.9
srcversion:     140276773A3090F6F33891F
depends:        
vermagic:       3.10.0-1160.21.1.el7.x86_64 SMP mod_unload modversions

# lsmod | grep main
main                   12426  0

# cat /proc/modules | grep '('
main 12426 0 - Live 0xffffffffc06c6000 (POE)

# dmesg
<<snip>>
[  797.912337] main: module verification failed: signature and/or required key missing - tainting kernel
[  797.912442] Hello world 1.

# convert2rhel
<<snip>>
[03/22/2021 09:55:11] TASK - [Convert: Final system checks before main conversion] **************
main 12426 0 - Live 0xffffffffc06c6000 (POE)
CRITICAL - Unsigned kernel module(s) has been detected. Third-party components are not supported per our software support policy
https://access.redhat.com/third-party-software-support.

 Uninstall or disable the module(s) and run convert2rhel again to continue with the conversion.
WARNING - Abnormal exit! Performing rollback ...

# rmmod main
# dmesg
<<snip>>
[ 1876.068697] Goodbye world 1.

# cp main.ko /lib/modules/3.10.0-1160.el7.x86_64/kernel/kernel
# convert2rhel
<<snip>>
CRITICAL - The following kernel module(s) are not supported in RHEL:
 kernel/kernel/main.ko
WARNING - Abnormal exit! Performing rollback ...
```

**Comment by shaded-enmity** [comment]
Same test on CentOS 8.3:
```
# cat main.c
#include <linux/module.h>	/* Needed by all modules */
#include <linux/kernel.h>	/* Needed for KERN_INFO */

int init_module(void)
{
	printk(KERN_INFO "Hello world 1.\n");
	return 0;
}

void cleanup_module(void)
{
	printk(KERN_INFO "Goodbye world 1.\n");
}
# ls -lah
-rw-r--r--.  1 root root  256 Mar 22 10:38 main.c
-rw-r--r--.  1 root root 211K Mar 22 10:39 main.ko
-rw-r--r--.  1 root root  826 Mar 22 10:39 main.mod.c
-rw-r--r--.  1 root root 109K Mar 22 10:39 main.mod.o
-rw-r--r--.  1 root root 104K Mar 22 10:39 main.o
-rw-r--r--.  1 root root  154 Mar 22 10:38 Makefile
-rw-r--r--.  1 root root   21 Mar 22 10:39 modules.order
-rw-r--r--.  1 root root    0 Mar 22 10:39 Module.symvers
# modinfo main.ko
filename:       /root/main.ko
rhelversion:    8.3
srcversion:     140276773A3090F6F33891F
depends:        
name:           main
vermagic:       4.18.0-240.15.1.el8_3.x86_64 SMP mod_unload modversions 
# insmod main.ko
# cat /proc/modules | grep "("
main 16384 0 - Live 0xffffffffc03f3000 (POE)
# dmesg
<<snip>>
[  690.281620] main: module verification failed: signature and/or required key missing - tainting kernel
[  690.281702] Hello world 1.
# convert2rhel
[03/22/2021 10:50:28] TASK - [Convert: Final system checks before main conversion] **************
main 16384 0 - Live 0xffffffffc03f3000 (POE)
CRITICAL - Unsigned kernel module(s) has been detected. Third-party components are not supported per our software support policy
https://access.redhat.com/third-party-software-support.

 Uninstall or disable the module(s) and run convert2rhel again to continue with the conversion.
WARNING - Abnormal exit! Performing rollback ...
# rmmod main.ko

# cp main.ko /lib/modules/4.18.0-240.15.1.el8_3.x86_64/kernel/kernel/
# convert2rhel
<<snip>>
CRITICAL - The following kernel module(s) are not supported in RHEL:
 kernel/kernel/main.ko
WARNING - Abnormal exit! Performing rollback ...

```

File: convert2rhel/checks.py
Comment (shaded-enmity) [inline convert2rhel/checks.py:201]: This currently prints path relative to `/lib/modules/$(uname -r)/` which can be confusing, we should print absolute path here.

File: convert2rhel/checks.py
Comment (shaded-enmity) [inline convert2rhel/checks.py:190]: 1. The output of grep is visible in stdout
2. It's not clear which kmod is the problem

```suggestion
    unsigned_modules, _ = run_subprocess("grep '(' /proc/modules")
    module_names = "\n  ".join([mod.split(' ')[0] for mod in unsigned_modules.splitlines()])
    if unsigned_modules:
        logger.critical(
            "Unsigned kernel module(s) has been detected. "
            "Third-party components are not supported per our "
            "software support policy\n%s.\n\n"
            "Uninstall or disable the following module(s) and run convert2rhel "
            "again to continue with the conversion:\n  %s",
            LINK_KMODS_RH_POLICY,
            module_names
        )
```

**Review by shaded-enmity — CHANGES_REQUESTED** [review]
Added 2 comments to make the output more clear about locations & names. Other than that my tests are passing and code in general LGTM.

File: convert2rhel/checks.py
Comment (bocekm) [inline convert2rhel/checks.py:79]:
```suggestion
        kernel_version, ret_code = utils.run_subprocess("uname -r")
        assert ret_code == 0
        kmod_str, ret_code = utils.run_subprocess("find /lib/modules/" + kernel_version.rstrip() + " -name *.ko*")
        assert kmod_str
        assert ret_code == 0
```

It works for me using the `run_subprocess` on CentOS 6/7/8.

**Review by bocekm — CHANGES_REQUESTED** [review]

File: convert2rhel/checks.py
Comment (SpyTec) [reply convert2rhel/checks.py:190]: Updated your suggestion @shaded-enmity to remove a dot after the URL. Some resolve the URL with the dot in it

So it will now be

```
CRITICAL - Unsigned kernel module(s) has been detected. Third-party components are not supported per our software support policy
https://access.redhat.com/third-party-software-support
```

instead of

```
CRITICAL - Unsigned kernel module(s) has been detected. Third-party components are not supported per our software support policy
https://access.redhat.com/third-party-software-support.
```

File: convert2rhel/checks.py
Comment (SpyTec) [reply convert2rhel/checks.py]: Right, I just don't know what it means

File: convert2rhel/unit_tests/checks_test.py
Comment (SpyTec) [reply convert2rhel/unit_tests/checks_test.py:146]: In this case we only have good and bad, so one if-else is enough. My suggestion is ultimately the same code, just shortened 

Can also add an explicit `elif host_kmods == HOST_MODULES_STUB_GOOD` in my suggestion too

File: convert2rhel/checks.py
Comment (bocekm) [reply convert2rhel/checks.py:79]: Calling `find` is not so pythonic. I'd prefer using for example `glob`, like [here](https://github.com/oamg/convert2rhel/blob/v0.18/convert2rhel/pkghandler.py#L728). But I'll not block the PR on that.

File: convert2rhel/checks.py
Comment (bocekm) [inline convert2rhel/checks.py:131]: I'm not in favor of code that is present only for the purpose of unit tests. I'm sure there's a way to write the tests to not require this workaround.

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py]: added

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py]: applied

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py:120]: me either but I was unable to find the way to solve it differently. The reason is such things like https://github.com/oamg/convert2rhel/blob/main/convert2rhel/systeminfo.py#L225. So the class initialized during the modules loading and  I can't affect it, without reinitialization. 

I believe it will be better to try to resolve this in a separate fix.

**Review by ZhukovGreen — COMMENTED** [review]
@bocekm done with fixing the review

**Review by bocekm — CHANGES_REQUESTED** [review]

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py:226]: There is a reason, that pre_ponr check needs subscription manger installed, because we are using repoquery

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py]: applied

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py]: renamed, added docstrings

**Review by ZhukovGreen — COMMENTED** [review]
@bocekm your turn:)

File: convert2rhel/checks.py
Comment (pirat89) [reply convert2rhel/checks.py:243]: In such a case, I suggest to mention it in the docstring and commit message so it's clear.

File: convert2rhel/checks.py
Comment (pirat89) [reply convert2rhel/checks.py:128]: @ZhukovGreen why it cannot reinitialized before the call of the function? To be honest I don't see the problem with that. what is preventing possibility to set the `system_info` in the test before you call the function?

Comment (shaded-enmity) [reply convert2rhel/checks.py:128]: DI or dependency inversion could be used here:
```python
def get_rhel_supported_kmods(system_info):
```

I.e. make it responsibility of the caller to provide `system_info`. Another benefit is that in testing you can supply the object without the need for mocking/monkeypatching.

**Review by shaded-enmity — COMMENTED** [review]
Generally LGTM, I will test again after 2pm today.

File: convert2rhel/checks.py
Comment (bocekm) [reply convert2rhel/checks.py:128]: This is essentially dead code - code that can never be executed at run-time.
After talking to @shaded-enmity, I agreed on not blocking the PR on things like this that do not directly affect the functionality. Here's a high priority task to remove the dead code and resolve the unit tests in a different way: https://issues.redhat.com/browse/OAMG-4634.

Comment (ZhukovGreen) [reply convert2rhel/checks.py]: As a DI it will work as well, but will be inconsistent, as this will be the only place where we would do this (at least now). @pirat89 advice works nicely, thx. OK, now we don't have this hack and it is moved to the fixture

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py:240]: We have docstrings for that:
```python
def perform_pre_checks():
    """Early checks after system facts should be added here."""


def perform_pre_ponr_checks():
    """Late checks before ponr should be added here."""
```
I believe it is clear enough, @pirat89 wdyt?

Comment (bocekm) [reply convert2rhel/checks.py:240]: I'd rename them first one to `perform_early_checks` to make it more descriptive.

File: convert2rhel/checks.py
Comment (bocekm) [inline convert2rhel/checks.py:127]: This would work only in case of using RHSM, not the custom repos.

```suggestion
        "--repoid " + repoid for repoid in get_enabled_rhel_repos()
```

Create the following new function in systeminfo.py:
```python
get_enabled_rhel_repos():
    """Return a list of repoids containing RHEL packages.

    These are either the repos enabled through RHSM or the custom repositories passed though CLI.
    """
    return system_info.submgr_enabled_repos if not tool_opts.disable_submgr else tool_opts.enablerepo
```
Use this new function also on two other places:
- https://github.com/oamg/convert2rhel/blob/v0.18/convert2rhel/pkghandler.py#L122
- https://github.com/oamg/convert2rhel/blob/v0.18/convert2rhel/pkghandler.py#L637

**Review by bocekm — CHANGES_REQUESTED** [review]

**Review by bocekm — CHANGES_REQUESTED** [review]

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py:142]: Added

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py:147]: All RHEL kmods packages starts with kernel* or kmod*. Added to docstring

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py:197]: I renamed it for consistency

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py]: thx

File: convert2rhel/checks.py
Comment (ZhukovGreen) [reply convert2rhel/checks.py:266]: ok, but this is out of scope

Comment (ZhukovGreen) [reply convert2rhel/checks.py:266]: will create a ticket for that

**Review by ZhukovGreen — COMMENTED** [review]
@bocekm done with the fixing review

**Review by shaded-enmity — APPROVED** [review]
Tested scenarios:

- CentOS 7 RHSM
- CentOS 7 Custom repositories
- CentOS 8 RHSM
- CentOS 8 Custom repositories

All tests were **green**. Approving.

**Review by bocekm — APPROVED** [review]

File: convert2rhel/checks.py
Comment (bocekm) [inline convert2rhel/checks.py:73]:
```suggestion
            ), print_output=False
```

File: convert2rhel/checks.py
Comment (bocekm) [inline convert2rhel/checks.py:160]: You need to handle the repoquery commands in a special way on CentOS/OL 8, otherwise you'll be getting:
```
Modular dependency problems:

 Problem 1: conflicting requests
  - nothing provides module(platform) needed by module ant:1.10:820181213135032:5ea3b708-0.x86_64
 Problem 2: conflicting requests
  - nothing provides module(platform:el8) needed by module container-tools:rhel8:8030020201124131330:830d479e-0.x86_64
```

For yum calls, we use the `--setopt=module_platform_id=platform:el8` option: https://github.com/oamg/convert2rhel/blob/main/convert2rhel/pkghandler.py#L113.

File: convert2rhel/checks.py
Comment (bocekm) [inline convert2rhel/checks.py:167]:
```suggestion
        ), print_output=False
```

**Review by bocekm — CHANGES_REQUESTED** [review]

**Review by ZhukovGreen — COMMENTED** [review]
@bocekm done

File: convert2rhel/checks.py
Comment (bocekm) [reply convert2rhel/checks.py]: https://issues.redhat.com/browse/OAMG-4668

**Review by bocekm — APPROVED** [review]

**Review by shaded-enmity — APPROVED** [review]
LGTM!

**Pull request merged**
